File: ex98.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    private boolean dfs(TreeNode root, long lo, long hi) {     
        // Not meet the requirement of a BST
        if (!(root.val < hi && root.val > lo))    return false;
        
        // Traverse to left and right children
        boolean left = (root.left != null) ? dfs(root.left, lo, root.val) : true;
        boolean right = (root.right != null) ? dfs(root.right, root.val, hi) : true;
        
        return left && right;
    }
    
    
    public boolean isValidBST(TreeNode root) {
        if (root == null)    return true;
        
        boolean left = (root.left != null) ? dfs(root.left, Long.MIN_VALUE, root.val) : true;        
        boolean right = (root.right != null) ? dfs(root.right, root.val, Long.MAX_VALUE) : true;     
        
        return left && right;
    }
}

File: code/ex98.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    bool res = 1;
    
public:
    bool isValidBST(TreeNode* root) {
        auto ret = dfs(root);
        return res;
    }
    
private:
    vector<long> dfs(TreeNode* root) {    // {min, max}
        if (root == NULL)    return {LONG_MAX, LONG_MIN};
        
        int val = root->val;
        auto left = dfs(root->left);
        auto right = dfs(root->right);
        
        if (left[1] >= val || right[0] <= val) {
            res = 0;
            return {LONG_MAX, LONG_MIN};
        }
        
        return {min((long)val, left[0]), max((long)val, right[1])};
    }
};